satpy/readers/yaml_reader.py

```diff
@@ -69,6 +69,9 @@ def _get_filebase(path, pattern):
     """Get the end of *path* of same length as *pattern*."""
     # convert any `/` on Windows to `\\`
     path = os.path.normpath(path)
+    # remove possible #mode=bytes URL suffix to support HTTP byte range
+    # requests for NetCDF
+    path = path.split('#')[0]
```

Comment:
Instead of splitting the string, I would consider using `urllib` (as someone else also mentioned in slack)
```python
import urllib
url = "https://mydomain.com/path/to/file.nc#mode=bytes"
url_components = urllib.parse.urlparse(url)
url_components.path  # == "/path/to/file.nc"
```
because you never know if the URL also contains other components like a query which will add a question mark or any other signs.
However, I am still not convinced that the yaml reader should have the responsibility of extracting the filename from the file path. Maybe the yaml reader should expect objects with a `.name` attribute so all the logic of extracting the filename is the responsibility of this object, which, if just looking for a `.name` attribute, could be any duck type of a `fileobj` or a `pathlib.Path`, or a `yarl.URL`, or ...

Reply:
I agree, just using split is a bit risky here.

Reply:
I like the idea of depending on `.path` but the main concern with this was speed. This method is called a lot and when given a lot of filenames (like using the MultiScene) it can add seconds to processing just to sort through the files. @gerritholl originally noticed this and a small change of where filename parsing happened cut off a lot of time iirc.

I'll try to look at it later.

Reply:
So the function I modified is indeed called a lot (inside a tight for loop), however, this operation to get `.path` would only have to be done *once* per filename. If force filenames to be convert to Path objects or some other "pre-processing" then I could see this `.path` stuff working out pretty well.

Reply:
@carloshorn I was playing around with this idea and realized one issue with the `Path` objects that I couldn't figure out. For some of our file patterns we need more than just the filename. Some readers encode the file type structure into directories so we really need a majority of the path. So for `pathlib.Path` the `.name` won't work. I would have all filenames parsed with urllib but it might be nice to duck type as you mentioned. Besides throwing in some if statements, any idea if there is a pathlib feature I'm missing here that could work for this? I suppose:

```
path = path_obj.path if hasattr(path_obj, 'path') else str(path_obj)
```

Reply:
Side note: the `#mode=bytes` is an anchor :facepalm: (a.k.a. a URL "fragment").

Reply:
> For some of our file patterns we need more than just the filename. Some readers encode the file type structure into directories so we really need a majority of the path.

That makes it of course a lot more complicated and makes it hard to use duck types which are not aware of a directory structure, e.g. file objects.
At the same time, I am wondering about this design choice. I do not know the details, and I guess there is a good reason for it. However, I have seen long discussions about filenames, and even if I do not see the same value in these names as others, I agree that it is a decent possibility to easily identify products, but it appears strange to me that a reader also constrains the user on how he/she needs to arrange his/her directory structure.
I guess the design choice traces back to the combination of file identifier and file location, but here an example could help to improve my understanding.
I also feel like I have not fully understood the function of the pattern. Is the pattern only used as a filter to determine which file can be read by the given reader? Or is the pattern also used to find appropriate files on the file system? Or is the pattern also used for the mapping from filename to reader (if yes, what is the tiebreaker if a filename matches the pattern of multiple readers?)?

Still, I do not see a problem with duck types. If we implement the duck inspection and let the readers, that put more constraints on their ducks, raise exceptions in case of missing attributes, it would not break any workflow.

> [...] any idea if there is a pathlib feature I'm missing here [...]

A `pathlib.Path` will not be a problem, because it is very aware of the directory structure. Beside the `.name` attribute, it also provides the `.parent` for the directory, `.parents` for direct access to other levels in the directory hierarchy, and `.parts` to split the path into directory names...

> I would have all filenames parsed with urllib [...]

If it is a string, we can parse it to get an object providing our desired attributes (where I personally prefer `yarl` over `urllib`), otherwise, we just pass whatever object was given to the reader and expect/hope that our users know what they are doing.

I guess that most reader patterns only constrain the base name of a given file path, so in most cases the `.name` attribute would be sufficient for a duck. If more is needed, the `.path` or `.parts` attribute could be the next thing to look for.

What do you think?

Reply:
> I also feel like I have not fully understood the function of the pattern.

- the patterns may contain metadata that might not be available in the data it self (for example timestamps for image formats), but are essential for the file handlers / readers
- the pattern _can_ be used to select the reader by matching the configured patterns with the filenames. User can override this by defining the reader in `Scene` init
- the patterns can be used to group the files together, e.g. for the use of `MultiScene()`

AFAIK the first one is the only that can't be overridden by the user with keyword arguments or sorting the files themselves.

Reply:
Many thanks @pnuu for the clarification. 

> AFAIK the first one is the only that can't be overridden by the user with keyword arguments or sorting the files themselves.

So the filename constraining pattern is a result of a lack of information, where technically speaking the pattern matching serves as an interface to access some missing metadata. This metadata is later on used for grouping, sorting, and maybe also as Scene attribute. So a possible solution would be, if we provide an alternative interface to pass this metadata and do not rely on the filename containing it. 
This is basically my original claim that it would be better to separate file location from file identifier, where now I would say that the file identifier is the collection of metadata that is somehow encrypted in the pattern.
Thinking about an implementation, there should be an object containing this required metadata for grouping etc, where one way of constructing it should be the extraction from the pattern. However, there should also be an alternative constructor which enables the user to pass this information from another source.

Reply:
Thinking it further, the yaml file for a reader, should contain the list of required metadata that it currently expects to be able to extract from the filename, so a user knows in advance what he/she needs to provide in order to use this reader with an alternative file identification interface.

Reply:
Filename matching has been a sore spot but I can't think of a better solution. There is some metadata in the filenames that is just "nice to have" but provides a better picture of the data being loaded. There is some that is necessary and not available in the file contents. However, the main point here is that it is needed because there is no better *easy* way to sort files for what python code should parse them. You either:

1. Sort by filename and give the filename to a class to read the files (what we do now).
2. Open the file, check some global attribute(s), then use that information to decide how the file should be read. *BUT* that requires knowing how to open the file in the first place. Not all files are netcdf or hdf5. Well then, you say, we could have code that looks at the extension of the file and determines if it knows how to read it. Then we are back to option 1 as some filenames don't have extensions or need to be decompressed in a certain way or sometimes they are even mis-extension'd.

We need to be able to determine what these filenames are *fast*. Having to open the files in any way isn't really an option. Having the user provide the information while possible, is not something non-programmer users can do. It also gets a little difficult when dealing with a lot of files. It ends up requiring the user to know low-level things about the files they are trying to read. At least I think so.

Reply:
To be clear, I'm not against trying something *not* filename based, but I have tried a lot of things in the past with my own projects (the ones that lead to Satpy). Everything other than filename based stuff seemed really hacky or didn't provide all of the features we currently have in satpy. If we take away the functionality that lets Satpy take *any* file and find a reader for it, then we have some more options. But this is also useful functionality to some people.

Reply:
@djhoese, please do not get me wrong, I am not against using the patterns, I just do not like that it constraints the data storage architecture. The current implementation only knows file systems, but imagine a situation where you are dealing with an object storage, which is designed to quickly provide the relevant metadata and a way to access the data itself. And even if you store the files on a file system, you might have a database containing the relevant metadata. In this case the pattern matching might be overhead, because it is certainly faster to query the database to get the metadata instead of decomposing a filename. For such use cases, it might be great to have an alternative way.
For a comprehensive implementation, it would be good to separate the extraction of required metadata by pattern matching from the subsequent sorting, grouping, ... operations, which I hope and guess that it can be done without adding too much overhead. In case the metadata is already present, it should be fine to skip this step.

> Having the user provide the information while possible, is not something non-programmer users can do.

I know that `satpy` aims at being a convenience product and I personally also appreciate it, but adding such an alternative metadata interface would not break any existing workflow. It only provides an interface for advanced use cases. 

I hope to find some time these days to present an implementation proposal.

Reply:
I like the idea (or the goal) of separating out the data source and getting the metadata from that data source. Your example of object storage is a little difficult because xarray and zarr are already able to read NetCDF-like data structures from remote object stores like S3 and they treat them like filenames. They load the initial metadata they need then when you request actual chunks of data additional requests are made to the remote storage. Something like a URL API could be a similar case though to what you are talking about where we have a URL with query parameters where data/metadata can be retrieved/accessed. There are more formal standards that the Pangeo group are dealing with more and more like STAC (https://stacspec.org/) that we could look in to leveraging, but at that point we should be using the Intake library (a pangeo project working towards better STAC support).

Maybe the main question is where do these responsibilities land? Who/what is doing these metadata extractions and data retrievals? How does the user interact with Satpy in these more complex cases? Does the user provide a file like object that provides properties to `.path` for sorting but also other base metadata (.start_time, etc.)? Does the user provide a "key" that Satpy readers should access from another provided "file system" (or "data store" or something)? Should readers (specifically the current "file handler" concept) be agnostic to where the data came from (general, yes, but some may be difficult)? Is the current file handler structure even flexible enough to handle this or should it be generalized to something like a "data repository" that provides datasets and satpy includes "dataset readers" instead of file handlers. This kind of goes along with xarray's idea of `.open_dataset("one_netcdf.nc")` returning an `xarray.Dataset` object.

:exploding_head: I should maybe wait until later on a Monday morning before thinking about these things.

Reply:
Have we compared the performance of `split` vs. `urllib`?